### PR TITLE
Archive attributes and runtime support for it

### PR DESCRIPTION
**Pull request by eeckstein**
Explanation: This is the second part for supporting NSKeyedArchives. It adds new attributes for classes (Doug's changes) and runtime support for it (Erik's changes).

Scope: This is a new feature

Issue: rdar://problem/32083946

Risk: Low. It only affects swift code which actually uses those attributes.

Testing: There are test case in this PR for swift CI

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@jrose-apple Can you please review this?

File: stdlib/public/runtime/RuntimeEntrySymbols.cpp [hunk 1]
```diff
@@ -36,6 +36,7 @@
 // implementations.
 #define FOR_CONV_DefaultCC(...)
 #define FOR_CONV_C_CC(...)
+#define FOR_CONV_SwiftCC(...)
```
Comment (gparker42) [inline stdlib/public/runtime/RuntimeEntrySymbols.cpp]: This needs documentation. The other CCs are described in the nearby comments.

**Comment by jrose-apple** [comment]
Already reviewed Doug's changes. Your changes mostly looked good to me but I'm not an IRGen person; maybe @jckarter or @rjmccall can sign off on it?

**Comment by jrose-apple** [comment]
I *am* getting these warnings on Linux:

```
/home/jrose/public/swift/include/swift/Runtime/Metadata.h:3315:1: warning: unknown attribute 'swiftcall' ignored [-Wunknown-attributes]
SWIFT_CC(swift)
^
/home/jrose/public/swift/include/swift/Runtime/Config.h:102:22: note: expanded from macro 'SWIFT_CC'
#define SWIFT_CC(CC) SWIFT_CC_##CC
                     ^
<scratch space>:39:1: note: expanded from here
SWIFT_CC_swift
^
/home/jrose/public/swift/include/swift/Runtime/Config.h:109:39: note: expanded from macro 'SWIFT_CC_swift'
#define SWIFT_CC_swift __attribute__((swiftcall))
                                      ^
```

Presumably this will happen on any platform where we don't yet use the just-built Clang to build the runtime.

**Comment by jckarter** [comment]
What do we need the `@_staticInitializeObjCMetadata` attribute for? That seems like it could easily be abused.

**Comment by eeckstein** [comment]
This attribute is automatically generated. It's for classes where we just have to make sure that it's metadata is initialized (before any unarchiving). For example non-generic subclasses of generic classes

**Comment by jckarter** [comment]
Can we prevent people from spelling the attribute in source code?

**Comment by eeckstein** [comment]
> Can we prevent people from spelling the attribute in source code?

A question for @DougGregor

**Comment by jrose-apple** [comment]
We could make it RejectByParser instead of UserInaccessible.

**Comment by DougGregor** [comment]
Yeah, that's a fine idea. I'll do the `_RejectByParser` thing in my PR to update the Fix-Its.

**Comment by jckarter** [comment]
Sounds good. Runtime changes LGTM.

**Comment by DougGregor** [comment]
That PR is over at https://github.com/apple/swift/pull/9473

**Pull request merged**